### PR TITLE
Add Go hooks 

### PR DESCRIPTION
**Pull request by jasonodoom**
#  #

## 🗣 Description ##

This PR includes additions to our pre-commit config to make use of  Golang hooks from [TekWizely/pre-commit-golang](https://github.com/TekWizely/pre-commit-golang).

## 💭 Motivation and context ##

As mentioned in the [initial PR](https://github.com/cisagov/skeleton-golang-package/pull/2#discussion_r1184218451) which initiated the creation of the current in order to implement the Staticcheck linter and other useful Go hooks. 

Validating both our workflows and any custom Actions we use is a useful addition to our [pre-commit](https://pre-commit.com) configuration. 

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -31,35 +31,74 @@ repos:
 
   # Text file hooks
   - repo: https://github.com/igorshubovych/markdownlint-cli
-    rev: v0.33.0
+    rev: v0.34.0
```
Comment (jsf9k) [inline .pre-commit-config.yaml]: All the updates to versions of extant `pre-commit` hooks should go in a separate PR.  @mcdonnnj usually creates such a PR just before we release the Lineage Kraken.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -31,35 +31,74 @@ repos:
...
+    rev: v3.3.1
     hooks:
       - id: validate_manifest
 
+  # Go hooks
+  - repo: https://github.com/TekWizely/pre-commit-golang
+    rev: v1.0.0-rc.1
+    hooks:
+    # Style Checkers
+      - id: go-lint
```
Comment (jsf9k) [inline .pre-commit-config.yaml]: Please be consistent in your indentation:
```suggestion
      # Style Checkers
      - id: go-lint
```

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -31,35 +31,74 @@ repos:
...
+      - id: go-critic
+    # StaticCheck
```
Comment (jsf9k) [inline .pre-commit-config.yaml]: Please be consistent in your indentation:
```suggestion
      # StaticCheck
```

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -31,35 +31,74 @@ repos:
...
+      - id: go-staticcheck-mod
+      - id: go-staticcheck-pkg
+      - id: go-staticcheck-repo-mod
+      - id: go-staticcheck-repo-pkg
+     # Go Build
```
Comment (jsf9k) [inline .pre-commit-config.yaml]: Please be consistent in your indentation:
```suggestion
      # Go Build
```

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -31,35 +31,74 @@ repos:
...
+      - id: go-build-mod
+      - id: go-build-pkg
+      - id: go-build-repo-mod
+      - id: go-build-repo-pkg
+
+      # Go Mod Tidy
```
Comment (jsf9k) [inline .pre-commit-config.yaml]: Please be consistent about whether you separate these sections with a blank line or not:
```suggestion
      - id: go-build-repo-pkg
      # Go Mod Tidy
```

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -31,35 +31,74 @@ repos:
...
+      - id: go-mod-tidy
+      - id: go-mod-tidy-repo
+
```
Comment (jsf9k) [inline .pre-commit-config.yaml]:
```suggestion
```

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -31,35 +31,74 @@ repos:
...
+      # Go Test
+      - id: go-test-mod
+      - id: go-test-pkg
+      - id: go-test-repo-mod
+      - id: go-test-repo-pkg
+
```
Comment (jsf9k) [inline .pre-commit-config.yaml]:
```suggestion
```

**Review by jsf9k — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -31,35 +31,74 @@ repos:
...
+    rev: v1.0.0-rc.1
+    hooks:
+    # Style Checkers
+      - id: go-lint
+      - id: go-critic
+    # StaticCheck
+      - id: go-staticcheck-mod
+      - id: go-staticcheck-pkg
+      - id: go-staticcheck-repo-mod
+      - id: go-staticcheck-repo-pkg
```
Comment (jsf9k) [inline .pre-commit-config.yaml]: In many of these blocks similar hooks are repeated like they are here.  I'm not clear as to whether or not this is actually necessary.

**Review by jsf9k — COMMENTED** [review]
I have a question.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (jasonodoom) [reply .pre-commit-config.yaml]: Aaah, I see what happened. The Golang pre-commit's README detailed to use `master` as rev But when I ran locally pre-commit complained and recommended `pre-commit autoupdate` which updated the others as well. I'll undo those.

File: .pre-commit-config.yaml (hunk 8 above)
Comment (jasonodoom) [reply .pre-commit-config.yaml]: This is based on the sample config: https://github.com/TekWizely/pre-commit-golang/blob/master/sample-config.yaml Should I keep only one per section?

**Comment by jsf9k** [comment]
For future reference, if you go to the "Files changed" tab you can group suggestions together into batches.  For example, in this case that would have allowed you to have a single "Fix indentation" commit and a single "Fix formatting" commit.

File: .pre-commit-config.yaml (hunk 8 above)
Comment (jsf9k) [reply .pre-commit-config.yaml]: I'm not sure.  It's difficult to tell what the differences are based just on the README for that project.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (jasonodoom) [reply .pre-commit-config.yaml]: Please see commit https://github.com/cisagov/skeleton-generic/pull/133/commits/c47038b51d4b537b9bd64b0299c1b99ca0b6e455

File: .pre-commit-config.yaml (hunk 8 above)
Comment (jasonodoom) [reply .pre-commit-config.yaml]: Yes, it should be safe to remove those extras.  Please see commit f8179f1aa31be65b2c363fd86badd6bf7883a033

Comment (mcdonnnj) [reply .pre-commit-config.yaml]: The documentation for what the hooks do can be found in the hook definition file: [.pre-commit-hooks.yaml](https://github.com/TekWizely/pre-commit-golang/blob/master/.pre-commit-hooks.yaml). Some examples:
```yaml
# ==============================================================================
# go-build-mod
#   * Folder-Based
#   * Recursive
#   * Targets first parent folder with a go.mod file
#   * Executes if any .go files modified
#   * Executes if go.mod modified
# ==============================================================================

# ==============================================================================
# go-build-pkg
#   * Folder-Based
#   * Targets folder containing staged file
#   * Executes if any .go files modified
# ==============================================================================

# ==============================================================================
# go-build-repo-mod
#   * Repo-Based
#   * Recursive
#   * Targets ALL folders with a go.mod file
#   * Executes if any .go files modified
#   * Executes if go.mod modified
# ==============================================================================

# ==============================================================================
# go-build-repo-pkg
#   * Repo-Based
#   * Recursive
#   * Executes if any .go files modified
# ==============================================================================
```

Comment (jsf9k) [reply .pre-commit-config.yaml]: @jasonodoom, how did you decide that `go-build-mod` was the right choice over, say, `go-build-repo-mod`.  The latter seems more inclusive to me.

Comment (jasonodoom) [reply .pre-commit-config.yaml]: @jsf9k yes was making these changes before I saw your comment based on Nick sharing the location of the hook definitions. I replaced the file recursive with those that have repo based options.

**Review by jsf9k — APPROVED** [review]

**Review by dav3r — APPROVED** [review]
👍 👍 

Friendly reminder to NOT merge this PR until we are ready for the next Kraken to begin.

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -60,6 +60,26 @@ repos:
     hooks:
       - id: validate_manifest
 
+  # Go hooks
+  - repo: https://github.com/TekWizely/pre-commit-golang
+    rev: v1.0.0-rc.1
+    hooks:
+      # Style Checkers
+      - id: go-lint
```
Comment (mcdonnnj) [inline .pre-commit-config.yaml]: Based on the status of https://github.com/golang/lint (archived and the note at the top of the README) supplemented with https://github.com/golang/go/issues/38968 should we remove this hook? I saw that some people switched to [golangci-lint](https://github.com/golangci/golangci-lint) (in addition or instead of the recommended alternatives in the README).

Comment (mcdonnnj) [reply .pre-commit-config.yaml]: @cisagov/team-ois

Comment (jsf9k) [reply .pre-commit-config.yaml]: Good point.  We should strongly consider removing it, and hopefully we can replace it with something else that continues to be maintained.  That said, I'd be OK with allowing it for now provided an issue was created to find a replacement. Presumably such a replacement will surface eventually, if it hasn't already.

Comment (mcdonnnj) [reply .pre-commit-config.yaml]: I ran into this when running this configuration locally. That was the only tool in the configuration that was neither already installed nor readily available from `brew`. So using this configuration for local development may be annoying as it stands.

Comment (jsf9k) [reply .pre-commit-config.yaml]: OK, go ahead and remove it with my blessing.

Comment (dav3r) [reply .pre-commit-config.yaml]: I'm ok removing golang/lint for now as well.  Someone can create another PR to propose a suitable replacement- no need to hold up progress any further for this.

Comment (mcdonnnj) [reply .pre-commit-config.yaml:68]: @jasonodoom

**Review by dav3r — APPROVED** [review]
👍
